src/stepFunctions/asl/asl-yaml-languageservice.ts

```diff
@@ -0,0 +1,245 @@
+
+        const aslCompletions = doCompleteAsl(document, position, currentDoc, yamlCompletions, {
+            ignoreColonOffset: true,
+        })
+
+        aslCompletions.items.forEach(completion => {
+            // format json completions for yaml
+            if (completion.textEdit) {
+                // textEdit can't be done on white-space so insert text is used instead
+                if (atSpace) {
+                    completion.insertText = completion.textEdit.newText
+                    // remove any commas from json-completions
+                    completion.insertText = completion.textEdit.newText.replace(/[\,]/g, '')
+                    completion.textEdit = undefined
+                } else {
+                    completion.textEdit.range.start.character = position.character
+                }
+            } else {
+                // remove null value auto-completions for 'string' nodes that don't support null values
+                completion.insertText = completion?.insertText?.replace(/\$\{1:null\}/g, '')
+            }
+        })
+        return Promise.resolve(aslCompletions)
+    }
+
+    languageService.doResolve = function(item: CompletionItem): Thenable<CompletionItem> {
+        return aslLanguageService.doResolve(item)
+    }
+
+    languageService.doHover = function(
+        document: TextDocument,
+        position: Position,
+        jsonDocument: JSONDocument
+    ): Thenable<Hover | null> {
+        const doc = parseYAML(document.getText())
+        const offset = document.offsetAt(position)
+        const currentDoc = matchOffsetToDocument(offset, doc)
+        if (currentDoc === null) {
+            return null!
+        }
+
+        const currentDocIndex = doc.documents.indexOf(currentDoc)
+        currentDoc.currentDocIndex = currentDocIndex
+
+        return aslLanguageService.doHover(document, position, currentDoc)
+    }
+
+    languageService.format = function(document: TextDocument, range: Range, options: FormattingOptions): TextEdit[] {
+        try {
+            const text = document.getText()
+
+            const formatted = prettier.format(text, Object.assign(options, { parser: 'yaml' }))
+
+            return [TextEdit.replace(Range.create(Position.create(0, 0), document.positionAt(text.length)), formatted)]
+        } catch (error) {
+            return []
+        }
+    }
+
+    languageService.findDocumentSymbols = function(document: TextDocument): SymbolInformation[] {
+        const doc = parseYAML(document.getText())
+        if (!doc || doc['documents'].length === 0) {
+            return []
+        }
+
+        let results: any[] = []
+        for (const yamlDoc of doc['documents']) {
+            if (yamlDoc.root) {
+                results = results.concat(aslLanguageService.findDocumentSymbols(document, yamlDoc))
+            }
+        }
+
+        return results
+    }
+
+    languageService.findDocumentSymbols2 = function(document: TextDocument): DocumentSymbol[] {
```

Comment:
It uses a findDocumentSymbols2 from the asl service instead of findDocumentSymbols. Asl service gets the findDocumentSymbols2 and findDocumentSymbols functions from the vscode-json-languageservice where there are minor differences between the two.

Reply:
> Could the two functions be generalized (passing in the appropriate `aslLanguageService.xxx` call) so that both functions do not have to be maintained in sync

I don't think so, this language service is just a wrapper for the asl language service which uses the two separate functions model so that could be an issue to bring up on that repo

Reply:
`findDocumentSymbols2` is part of the `LanguageService` interface from vscode-json-languageservice ([link](https://github.com/microsoft/vscode-json-languageservice/blob/master/src/jsonLanguageService.ts#L46)) which returns `DocumentSymbol` instead of `SymbolInformation`.

As far as I can tell, we only use it [here](https://github.com/aws/aws-toolkit-vscode/blob/ab650f54d0a4526aadcbff2979136f27ba735717/src/stepFunctions/asl/aslServer.ts#L388) in `aslServer.ts`.